src/ray/util/io_service_pool.h

```diff
@@ -0,0 +1,65 @@
+#ifndef RAY_UTIL_IO_SERVICE_POOL_H
+#define RAY_UTIL_IO_SERVICE_POOL_H
+
+#include <atomic>
+#include <boost/asio.hpp>
+#include <thread>
+
+namespace ray {
+
+/// \class IOServicePool
+/// The io_service pool. Each io_service owns a thread.
+/// To get io_service from this pool should call `Run()` first.
+/// Before exit, `Stop()` must be called.
+class IOServicePool {
+ public:
+  IOServicePool(size_t io_service_num);
+
+  ~IOServicePool();
+
+  void Run();
+
+  void Stop();
+
+  /// Select io_service by round robin.
+  ///
+  /// \return io_service
+  boost::asio::io_service *Get();
+
+  /// Select io_service by hash.
+  ///
+  /// \param hash Use this hash to pick a io_service.
+  /// The same hash will alway get the same io_service.
+  /// \return io_service
+  boost::asio::io_service *Get(size_t hash);
+
+  /// Get all io_service.
+  /// This is only use for RedisClient::Connect().
+  std::vector<boost::asio::io_service *> GetAll();
+
+ private:
+  size_t io_service_num_{0};
+
+  std::vector<std::thread *> threads_;
```

Comment:
Could we use `unique_ptr` or the like here to avoid manually managing these with `delete` etc.?

Reply:
Both way is ok.

Reply:
Can we switch it to a `unique_ptr` though (in addition to any other instances in the PR where this is done)? in addition to being more in line with modern C++ practice, a `unique_ptr` actually indicates ownership. A raw pointer is generally reserved for non-owning pointers, so it's misleading to see it here.

Reply:
Raw pointer also used when we know clearly the owner of it and the lifetime of it.

Reply:
Would you mind changing it to `unique_ptr`? It's standard practice, and there's no benefit to doing it this way, only potential confusion.

Reply:
I agree it's better to use unique_ptr here, in order to avoid confusions and issues if others change it in the future.

Reply:
Oh, I just realized we might not even need the pointer by the way—I haven't checked in detail. If it's not necessary, we should just do `std::vector<std::thread>` with `std::move` as appropriate.

Reply:
Still both way is ok. If we know clearly the ownership and the lifetime of the pointer, raw pointer is ok.

Both thread and io_service can be object instead of pointer. But it's convenience for "GetAll" if we use pointer.

Reply:
@micafan: What is the benefit of avoiding `unique_ptr` though? I don't understand.

Reply:
Using raw pointers to indicate ownership is **not** ok. We should change this.